### PR TITLE
ci: move benchmark into Github Actions

### PR DESCRIPTION
**Pull request by yihau**
#### Problem

we don't store those benchmark results properly and it's hard to compare with previous runs.

#### Summary of Changes

- try to use `benchmark-action/github-action-benchmark@v1`. it will store result and generate a useful github pages. it looks like https://yihau.github.io/solana/dev/bench/
- remove some useless pipeline scripts
- update pipelines

I would like to use this one as the first version and will add more missing benchmarks into it in the following PRs.

--- 
before merge

- [x] create a gh-page branch
- [x] set up self-hosted runner

### CODE REVIEW

File: .github/workflows/bench.yml [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    name: rust benchmark
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        pkg:
+          - solana-runtime
+          - solana-gossip
+          - solana-poh
+          - solana-sdk
```
Comment (brooksprumo) [inline .github/workflows/bench.yml]: Can accounts-db be benched too?

```suggestion
        pkg:
          - solana-runtime
          - solana-gossip
          - solana-poh
          - solana-sdk
          - solana-accounts-db
```

File: Cargo.toml [hunk 2]
```diff
@@ -3,6 +3,9 @@ inherits = "release"
 debug = true
 split-debuginfo = "packed"
 
+[profile.bench]
+lto = true
+
```
Comment (brooksprumo) [inline Cargo.toml]: Why use lto here just for the benches?

File: .github/workflows/bench.yml (hunk 1 above)
Comment (yihau) [reply .github/workflows/bench.yml]: sure! added!

File: Cargo.toml (hunk 2 above)
Comment (yihau) [reply Cargo.toml]: I copied the example config directly 🙈 I have removed it. I think we can add this one back when we publish things with the same arg.

**Comment by yihau** [comment]
@brooksprumo sorry for the late response. this one is ready for another look!

**Review by brooksprumo — COMMENTED** [review]
The changes look good to me. Note however that I don't know a lot about CI/build pipelines.

Will this guarantee the same machine(s) run the benchmarks each time? If we want to have a maximally useful regression suite, we'll need to run the benchmarks on the same machine(s) every time. And also ensure nothing else is running on the machines too.

**Comment by yihau** [comment]
yes! I only set up 1 machine for running benchmark. it means it will run on the same machine every time!

**Comment by yihau** [comment]
details for this PR:

- I only set up one server for benchmark atm. we can ensure we run the same benchmark tests on the same server and that no other tests are running simultaneously.
- the data will be stored/updated on gh-pages branch. the file is `dev/bench/data.js`. use my fork as example: https://github.com/yihau/solana/blob/gh-pages/dev/bench/data.js and the page looks like: https://yihau.github.io/solana/dev/bench
- I set a limit of 100 for each chart so that it won't looks too dense: https://github.com/anza-xyz/agave/pull/2064/files#diff-a5e740be96415373789689f814583e93ff2a8f05eae6481e94505fd6cb6bc6a7R71. (their example page: https://benchmark-action.github.io/github-action-benchmark/dev/bench/)
- I don't stop the previous run when a new one comes. would like to keep every result but it also means we might not see the results in time if our repository is busy.
- our page should be hosted at https://anza-xyz.github.io/agave/dev/bench when this one gets merged
- there is a potential bug where if github schedule ci jobs out of order, our chart will looks a little weird. there may be some peaks or troughs. I don't have a solution for it, but I guess it should be fine since if something really happen, we should see a similar trend in subsequent commits.
- it will take ~26m for current pipeline: https://github.com/yihau/solana/actions/runs/9926570429

**Comment by brooksprumo** [comment]
Some of the accounts-db benchmarks use Criterion, and I don't see them in the example output. It looks like the github example page *does* show criterion results.

Searching, I found we likely need to do something like what is here:
https://github.com/benchmark-action/github-action-benchmark/blob/fe4e90e7735310d03ca4b1d8f0eb1d425c55be00/examples/criterion-rs/README.md?plain=1

Is that possible?

**Comment by yihau** [comment]
@brooksprumo yeah should be possible but I will try to another plan for the benchmark haha.

**Comment by yihau** [comment]
I will try to store benchmark data into influxdb so that we can do more thing. will open another PR for it!

**Pull request closed without merging**